cmake/Googletest.cmake

```diff
@@ -31,18 +31,14 @@ macro(fetch_googletest)
         endif()
 
         message(STATUS "========== Installing GoogleTest... ==========")
-        set(FORCE_SHARED_CRT "")
-        if(IS_WINDOWS)
-            set(FORCE_SHARED_CRT -DFORCE_SHARED_CRT=OFF)
-        endif()
-        execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM} . ${FORCE_SHARED_CRT}
+           execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM} .
```

Comment:
CMAKE_MAKE_PROGRAM can also be a path, meaning it can also have spaces, meaning this arg should also be in quotes!

Reply:
also - I just noticed that the indentation is wonky

Reply:
> 
> 
> CMAKE_MAKE_PROGRAM can also be a path, meaning it can also have spaces, meaning this arg should also be in quotes!

@elrond79   We really don't want to do that! I have seen this issues so many times before and had my reasons to not wrap this around quotes.   You don't want to double quote things. :)
```
Access is denied
CMake Error: Generator: execution of make failed. Make command was: ""C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/Bin/MSBuild.exe"" "ALL_BUILD.vcxproj" "/p:Configuration=RelWithDebInfo" "/p:Platform=x64" "/p:VisualStudioVersion=15.0"
CMake Error at cmake/Googletest.cmake:48 (message):
```
This whole quote, white space stuff IMHO is really annoying.....  Something for CMake folks to think about.

> Also - I was going to add a comment that you should avoid commit messages like "final touches" - both because it gives no useful information in the future, when you're scanning through the history, but also because they often end up not being final. And I guess this review provides a good example of that. =) (In my experience, naming something "final" is almost a jinx!)

We will have a separate PR soon for code quality/styling ,etc.... That includes both C++ and CMake files. Can we please wait for that? It is just a waste of time and energy now to do this minor adjustments in this PR.

Reply:
This happens because you inserted the quote at the wrong place - in the middle of the argument, after the equals, like this:

`-DCMAKE_MAKE_PROGRAM="${CMAKE_MAKE_PROGRAM}"`

...instead of at the start / end of the argument, like this:

`"-DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}"`

I just tested, and confirmed that if I put the quote in the middle, I get the error you encountered, but if you put it at the start, it works correctly.

I agree that dealing with quoting / whitespace issues in cmake is a pain - but it's not going to get better.  Cmake is one of those languages (like tcl, or bash) where there's only one variable type - string.  Those languages invariably make quoting / whitespace / etc a royal pain.  It's probably the number one thing I dislike about cmake.

If you're curious, the exact rules governing this particular behavior can be found here:

https://cmake.org/cmake/help/v3.12/manual/cmake-language.7.html#unquoted-argument

...specifically, in the "Note: To support legacy CMake code.." section.

As for the whitespace - yes, we're planning on introducing a formal coding style.  However, I don't know of ANY cmake coding style / convention where the indentation as it currently is would be correct.  And the old code - before this PR - was correct.

Also, I get that you're fatigued by this process - I am too - but the best way to deal with long reviews like this is to try to make shorter, individual PRs in the future, not compromise on code quality.

Reply:
> This happens because you inserted the quote at the wrong place - in the middle of the argument, after the equals, like this.
> I just tested, and confirmed that if I put the quote in the middle, I get the error you encountered, but if you put it at the start, it works correctly.

I can't possibly go through every permutations of placing quotes before and after variables and make a call if it is right or wrong. I am honestly lost for words. Let's go with we have for now.

> I agree that dealing with quoting / whitespace issues in cmake is a pain - but it's not going to get better. Cmake is one of those languages (like tcl, or bash) where there's only one variable type - string. Those languages invariably make quoting / whitespace / etc a royal pain. It's probably the number one thing I dislike about cmake.

There are many aspects of CMake that I really like, but I do agree with you on this. 

> As for the whitespace - yes, we're planning on introducing a formal coding style. However, I don't know of ANY cmake coding style / convention where the indentation as it currently is would be correct. And the old code - before this PR - was correct.

Me neither!! As long as there is consistency in coding style, naming conventions, etc.... that's all matters.

Reply:
> > This happens because you inserted the quote at the wrong place - in the middle of the argument, after the equals, like this.
> > I just tested, and confirmed that if I put the quote in the middle, I get the error you encountered, but if you put it at the start, it works correctly.
> 
> I can't possibly go through every permutations of placing quotes before and after variables and make a call if it is right or wrong. I am honestly lost for words. Let's go with we have for now.

I certainly don’t expect you to randomly try permutations of of where to place quotes. I’m telling you the correct way to quote an argument and I’ve linked you to the documentation.  If you're confused, please take the time to understand the proper syntax so that we can merge this efficiently.

If you have an argument to make for why the code as stands is superior - or even equivalent - to what I’ve proposed, I will listen. Otherwise, please make the requested change.

Reply:
> I certainly don’t expect you to randomly try permutations of of where to place quotes. I’m telling you the correct way to quote an argument and I’ve linked you to the documentation. If you're confused, please take the time to understand the proper syntax so that we can merge this efficiently. If you have an argument to make for why the code as stands is superior - or even equivalent - to what I’ve proposed, I will listen. Otherwise, please make the requested change.

I believe you've misunderstood my previous comment and it is completely unnecessary to be so defensive and reactive. 

Sorry but I don't really understand what you mean by code superiority?

No, I am not confused. As I mentioned in my previous comment, when I wrap ${CMAKE_MAKE_PROGRAM} around quotes ( the wrong way apparently and thanks for correcting ), cmake build fails and from the error message it is apparent that the CMAKE_MAKE_PROGRAM has already have quote around it:

`""C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/Bin/MSBuild.exe""`
 
This tells me that we really don't need the extra quotes around arguments being passed to the `execute_process` command. I can't also see anything being mentioned in the official cmake doc:

https://cmake.org/cmake/help/v3.12/command/execute_process.html

Fyi, the path to visual studio on Windows has many spaces and special characters and to this day I haven't seen a single a issue being reported around this. Here is another example from Kitware, and I still don't see any extra quotes around any of the arguments:

https://github.com/Kitware/CMake/blob/1fb483922557003aedf59b01742928543977e013/Tests/FindPackageModeMakefileTest/CMakeLists.txt#L9

Even official google cmake example doesn't have quotes around things:
https://github.com/google/googletest/blob/master/googletest/README.md

Also Fyi, I was planning to replace all these codes with `FetchContent` module which is a more elegant way of building third-party source codes but don't believe it is a right time to do it:
https://cmake.org/cmake/help/v3.11/module/FetchContent.html
https://cliutils.gitlab.io/modern-cmake/chapters/projects/fetch.html

Reply:
> As I mentioned in my previous comment, when I wrap ${CMAKE_MAKE_PROGRAM} around quotes ( the wrong way apparently and thanks for correcting ), cmake build fails and from the error message it is apparent that the CMAKE_MAKE_PROGRAM has already have quote around it:
> 
> `""C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/Bin/MSBuild.exe""`
> 

Actually, the command it tries to execute here only has one set of quotes - the ones introduced by inserting quotes in the middle of a cmake arg.  The extra set is just introduced by cmake's error printing - it's doing the equivalent of:

```
message(STATUS "Make command was: \"${arg1}\" \"${arg2}\ ..."")
```

You can confirm this by adding a line like:

```
message(STATUS "CMAKE_MAKE_PROGRAM: --${CMAKE_MAKE_PROGRAM}--")
```

Or, if you're worried `execute_process` is somehow doing something special, by doing:

```
execute_process(COMMAND "/usr/bin/python" -c [[
from __future__ import print_function
import sys

print("args (num: {})".format(len(sys.argv)))
for i, arg in enumerate(sys.argv):
    print("{}: {}".format(i, arg))
]] foo bar "-DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}")
```

So, there are no extra quotes, even if `${CMAKE_MAKE_PROGRAM}` contains spaces, which I've confirmed.

> Fyi, the path to visual studio on Windows has many spaces and special characters and to this day I haven't seen a single a issue being reported around this. Here is another example from Kitware, and I still don't see any extra quotes around any of the arguments:
> 
> https://github.com/Kitware/CMake/blob/1fb483922557003aedf59b01742928543977e013/Tests/FindPackageModeMakefileTest/CMakeLists.txt#L9
> 
> Even official google cmake example doesn't have quotes around things:
> https://github.com/google/googletest/blob/master/googletest/README.md

These are valid arguments / points!  While I noticed the "double quotes" in the error printout you made, I admit I overlooked that there were spaces already present in the path the followed.

This led me to question why unquoted expansion of a var was spaces was working.  I had a long explanation here for why that was, but I don't know if you're interested in understanding the mechanics, so I'll cut to the chase:

It turns out that my understandings] of how quote expansion worked, in relation to vars with spaces in them (and other "weird paths"), was wrong.  NONE of the args in this command need to be quoted in order to deal with spaces correctly.  I'm embarrassed I didn't already know this.

Given that, you've now convinced me that the version in https://github.com/Autodesk/maya-usd/pull/198/commits/96156e7ab02ca1a3f4a236b5ae6b9a8c5d305f81 is essentially equivalent to my proposed change.  I still feel that it would be better to be consistent, and either do what I suggested:

```
execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" "-DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}" .
```

...or leave it in it's original form:

```
execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}
```

...but if you're set on leaving this line as it is, I no longer object.

When investigating why this worked, however, I did come across one other line which DOES need to be fixed.  Before you protest - if you absolutely do not want to make any further changes to your PR, then I'll make a separate PR for it myself.

The line in question is [this one](https://github.com/Autodesk/maya-usd/pull/198#pullrequestreview-342588721):

```
install(CODE "execute_process(COMMAND \"${Python_EXECUTABLE}\" -m compileall ${AL_INSTALL_PREFIX} )")
```

...and the issue is that the `${AL_INSTALL_PREFIX}` should be quoted. This one definitely WILL fail if `${AL_INSTALL_PREFIX}` contains spaces.

If you're curious to know the details of how quoted / unquoted variable expansion works (and why the quotes are necessary here, but aren't above), let me know, and I'll gladly explain it.

Otherwise, let me know whether you'll make this fix or want me to, and we'll move on.  Sound fair?

Reply:
> It turns out that my understandings] of how quote expansion worked, in relation to vars with spaces in them (and other "weird paths"), was wrong. NONE of the args in this command need to be quoted in order to deal with spaces correctly. I'm embarrassed I didn't already know this.

Same here! Glad we all learned something here.

> This led me to question why unquoted expansion of a var with spaces was working. I had a long explanation here for why that was, but I don't know if you're interested in understanding the mechanics. If you're curious to know the details of how quoted / unquoted variable expansion works (and why the quotes are necessary here, but aren't above), let me know, and I'll gladly explain what I learned.

I don't mind learning something new but let's put it off for another day!

Fyi, here is the C++ implementation of ExecuteProcessCommand which I haven't gotten the chance to fully read and should provide more clues.

https://github.com/Kitware/CMake/blob/master/Source/cmExecuteProcessCommand.cxx

> When investigating why this worked, however, I did come across one other line which DOES need to be fixed. Before you protest - if you absolutely do not want to make any further changes to your PR, then I'll make a separate PR for it myself.

Please go ahead and make the necessary changes!

Reply:
Actually, I forgot that I had push access to the repo, so I just made the change on this branch - which is simpler, since otherwise I'd to resolve conflicts, or make this PR a dependency, since it's changing the same line.  Please make sure to fetch and update!